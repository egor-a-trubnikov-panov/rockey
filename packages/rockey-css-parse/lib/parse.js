import cssToJSON from './cssToJSON';
import createMixin from './mixins/createMixin';
import interpolateWithMixins from './mixins/interpolateWithMixins';
import * as classnames from './utils/classnames';

const isMedia = key => key.indexOf('@media') === 0;
const isKeyFrames = key => key.indexOf('@keyframes') === 0;
const isNot = key => key.indexOf(':not') === 0;
const NOT_REGEX = /\(([^\)]+)\)(.*)/;

// @replace-start
const isModificatorStartSymbol = symbol => {
  return symbol === '@' || symbol === ':' || symbol === '[';
};
// @replace-end

// @replace-start
const isModificatorEndSymbol = symbol => {
  return symbol === '{' || symbol === ';';
};
// @replace-end

const isStartsWithModificator = raw => {
  return (
    0 === raw.indexOf('@media') ||
    0 === raw.indexOf('[') ||
    0 === raw.indexOf('@keyframes') ||
    0 === raw.indexOf('::placeholder') ||
    0 === raw.indexOf('::after') ||
    0 === raw.indexOf('::before') ||
    0 === raw.indexOf('::first-letter') ||
    0 === raw.indexOf('::first-line') ||
    0 === raw.indexOf(':active') ||
    0 === raw.indexOf(':checked') ||
    0 === raw.indexOf(':disabled') ||
    0 === raw.indexOf(':empty') ||
    0 === raw.indexOf(':enabled') ||
    0 === raw.indexOf(':first-child') ||
    0 === raw.indexOf(':first-of-type') ||
    0 === raw.indexOf(':focus') ||
    0 === raw.indexOf(':hover') ||
    0 === raw.indexOf(':in-range') ||
    0 === raw.indexOf(':invalid') ||
    0 === raw.indexOf(':lang') ||
    0 === raw.indexOf(':last-child') ||
    0 === raw.indexOf(':last-of-type') ||
    0 === raw.indexOf(':link') ||
    0 === raw.indexOf(':not') ||
    0 === raw.indexOf(':nth-child') ||
    0 === raw.indexOf(':nth-last-child') ||
    0 === raw.indexOf(':nth-last-of-type') ||
    0 === raw.indexOf(':nth-of-type') ||
    0 === raw.indexOf(':only-of-type') ||
    0 === raw.indexOf(':only-child') ||
    0 === raw.indexOf(':optional') ||
    0 === raw.indexOf(':out-of-range') ||
    0 === raw.indexOf(':read-only') ||
    0 === raw.indexOf(':read-write') ||
    0 === raw.indexOf(':required') ||
    0 === raw.indexOf(':root') ||
    0 === raw.indexOf(':target') ||
    0 === raw.indexOf(':valid') ||
    0 === raw.indexOf(':visited')
  );
};

const shouldGenerateSelectors = parent => parent.type !== 'keyframes';

export default function createParser(config = {}) {
  const plugins = config.plugins || [];

  const actions = {
    // getClassName: classnames.getClassName(config.getClassName),
    getMixinClassName: classnames.getMixinClassName(config.getMixinClassName),
    // getSelector: classnames.getSelector(config.getClassName),
    getAnimationName: config.getAnimationName
      ? config.getAnimationName
      : name => name,
  };

  const getComponentName = classnames.getComponentName;

  return function parse(strings, ...values) {
    const context = {
      hasAnimations: false,
      animations: {},
      classnames: {
        components: {},
        selector: {},
      },
    };

    const classname = classnames.getClassName(config.getClassName);
    const getClassName = component => {
      if (context.classnames.components[component]) {
        return context.classnames.components[component];
      }
      const className = classname(component);
      context.classnames.components[component] = className;

      return className;
    };

    const selector = classnames.getSelector(getClassName);
    const getSelector = component => {
      if (context.classnames.selector[component]) {
        return context.classnames.selector[component];
      }
      const className = selector(component);
      context.classnames.selector[component] = className;

      return className;
    };

    let precss = [];

    let raw = null;
    let mixinsFunctions = {};

    if (values.length === 0) {
      raw = Array.isArray(strings) ? strings[0] : strings;
    } else {
      const interpolated = interpolateWithMixins(strings, ...values);
      raw = interpolated.raw;
      mixinsFunctions = interpolated.mixinsFunctions;
    }

    const inline = raw.replace(/\r|\n/g, '').replace(/\s+/g, ' ');

    const root = toPreCSS(inline, {
      type: 'root',
      selector: [],
      media: null,
      root: [],
    });

    return {
      classList: root.reduce((components, c) => {
        components[getComponentName(c)] = getClassName(c);
        return components;
      }, {}),
      precss,
    };

    function toPreCSS(raw, parent = {}) {
      let openedBrackets = 0;
      let openedModificatorBrackets = 0;
      let styles = '';

      let rootComponents = [];
      let components = [];

      let possibleModificator = 0;
      let modificator = '';

      let current = '';
      let currentBackup = '';

      // @replace-start
      const restore = () => {
        current = currentBackup + current;
        currentBackup = '';
      };
      // @replace-end

      // @replace-start
      const clearAndRestore = () => {
        current = currentBackup;
        currentBackup = '';
      };
      // @replace-end

      // @replace-start
      const backup = () => {
        currentBackup = current;
        current = '';
      };
      // @replace-end

      // @replace-start
      const possibleModificatorStart = () => {
        possibleModificator = true;
      };
      // @replace-end

      // @replace-start
      const correctModificatorPosition = () => {
        return (s => !s.length || s.slice(-1) === ';' || s.slice(-1) === '{')(
          current.trim()
        );
      };
      // @replace-end

      // @replace-start
      const isPossibleModificator = () => {
        return possibleModificator;
      };
      // @replace-end

      // @replace-start
      const possibleModificatorEnd = () => {
        possibleModificator = false;
      };
      // @replace-end

      // @replace-start
      const isModificator = () => {
        return isStartsWithModificator(current.trim());
      };
      // @replace-end

      // @replace-start
      const openBracket = () => {
        openedBrackets++;
      };
      // @replace-end

      // @replace-start
      const closeBracket = () => {
        openedBrackets--;
      };
      // @replace-end

      // @replace-start
      const isAllBracketsClosed = () => {
        return openedBrackets === 0;
      };
      // @replace-end

      // @replace-start
      const openModificatorBracket = () => {
        openedModificatorBrackets++;
      };
      // @replace-end

      // @replace-start
      const closeModificatorBracket = () => {
        openedModificatorBrackets--;
      };
      // @replace-end

      // @replace-start
      const isAllModificatorBracketsClosed = () => {
        return openedModificatorBrackets === 0;
      };
      // @replace-end

      // @replace-start
      const startModificator = () => {
        modificator = current;
        current = '';
      };
      // @replace-end

      // @replace-start
      const saveModificator = () => {
        modificator = modificator.trim();

        let selector = [];
        const media = isMedia(modificator);
        const keyframes = isKeyFrames(modificator);

        if (shouldGenerateSelectors(parent)) {
          if (media || keyframes) {
            selector = [...parent.selector];
          } else {
            modificator.split(',').forEach(m => {
              m = m.trim();

              if (isNot(m)) {
                const matches = m.match(NOT_REGEX);
                m = `:not(${getSelector(matches[1])})${matches[2]}`;
              }

              if (parent.selector.length) {
                parent.selector.forEach(s => {
                  selector.push(s + m);
                });
              } else {
                selector.push(m);
              }
            });
          }
        }

        if (keyframes) {
          let animationName = modificator.split(' ')[1];

          let uniqAnimationName = actions.getAnimationName(animationName);

          context.hasAnimations = true;
          context.animations[animationName] = uniqAnimationName;

          const backup = precss;
          precss = [];

          toPreCSS(current, {
            type: 'keyframes',
            root: parent.root,
          });

          backup.push({
            media: media ? modificator : parent.media,
            selector: modificator.replace(animationName, uniqAnimationName),
            frames: precss,
          });

          precss = backup;
        } else {
          toPreCSS(current, {
            type: 'modificator',
            media: media ? modificator : parent.media,
            selector,
            root: parent.root,
          });
        }

        modificator = '';
      };
      // @replace-end

      // @replace-start
      const isInsideModificator = () => {
        return !!modificator;
      };
      // @replace-end

      // @replace-start
      const isRootComponentBlockStarts = symbol => {
        return symbol === '{' && isAllBracketsClosed();
      };
      // @replace-end

      // @replace-start
      const startComponent = () => {
        let i = null;
        for (i = current.length; i > 0; i--) {
          const symbol = current[i];
          if (symbol === ';') {
            i++;
            break;
          }
        }

        let currentComponents = [];

        if (i) {
          styles += current.slice(0, i);
          currentComponents = current.slice(i).split(',');
        } else {
          currentComponents = current.split(',');
        }

        currentComponents.forEach(c => {
          components.push(c.trim());
        });

        current = '';
      };
      // @replace-end

      // @replace-start
      const isInsideComponent = () => {
        return !!components.length;
      };
      // @replace-end

      // @replace-start
      const save = symbol => {
        current += symbol;
      };
      // @replace-end

      // @replace-start
      const saveComponent = () => {
        const selector = [];

        if (shouldGenerateSelectors(parent)) {
          components.forEach(c => {
            rootComponents.push(c);

            const className = getSelector(c);

            if (parent.selector.length) {
              parent.selector.forEach(p => {
                selector.push(`${p} ${className}`);
              });
            } else {
              selector.push(className);
            }
          });
        }

        toPreCSS(current, {
          type: 'component',
          media: parent.media,
          root: parent.root.length ? parent.root : components,
          selector: parent.type === 'keyframes' ? components : selector,
        });

        components = [];
        current = '';
      };
      // @replace-end

      // @replace-start
      const updateStylesTail = () => {
        if (current) {
          styles += current;
        }
      };
      // @replace-end

      // ---
      const mixins = [];
      let possibleMixin = false;
      let mixin = '';

      // @replace-start
      const isMixinStartSymbol = symbol => {
        return symbol === '_';
      };
      // @replace-end

      // @replace-start
      const isMixinEndSymbol = symbol => {
        return symbol === ';' || symbol === ' ';
      };
      // @replace-end

      // @replace-start
      const possibleMixinStart = () => {
        possibleMixin = true;
      };
      // @replace-end

      // @replace-start
      const possibleMixinEnd = () => {
        possibleMixin = false;
      };
      // @replace-end

      // @replace-start
      const isPossibleMixin = () => {
        return possibleMixin;
      };
      // @replace-end

      // @replace-start
      const saveMixinString = symbol => {
        mixin += symbol;
      };
      // @replace-end

      // @replace-start
      const clearMixin = () => {
        mixin = '';
      };
      // @replace-end

      // @replace-start
      const removeMixinFromStyles = () => {
        current = current.replace(mixin, '');
      };
      // @replace-end

      // @replace-start
      const saveMixin = () => {
        const mixinFunction = createMixin({
          className: actions.getMixinClassName(mixinsFunctions[mixin]),
          selector: parent.selector,
          root: parent.root,
          func: mixinsFunctions[mixin],
          parse,
        });

        mixins.push(mixinFunction);
      };
      // @replace-end
      let size = raw.length;
      for (let i = 0; i < size; i++) {
        const symbol = raw[i];

        if (!isInsideComponent() && !isInsideModificator()) {
          if (isMixinStartSymbol(symbol)) {
            possibleMixinStart();
          }

          if (isPossibleMixin()) {
            if (isMixinEndSymbol(symbol)) {
              possibleMixinEnd();
              saveMixin();
              removeMixinFromStyles();
              clearMixin();
              continue;
            } else {
              saveMixinString(symbol);
            }
          }

          if (
            !isPossibleModificator() &&
            isModificatorStartSymbol(symbol) &&
            correctModificatorPosition()
          ) {
            possibleModificatorStart();
            backup();
          }

          if (isPossibleModificator() && isModificatorEndSymbol(symbol)) {
            if (isModificator()) {
              startModificator();
            } else {
              restore();
            }

            possibleModificatorEnd();
          }
        }

        if (
          !isPossibleModificator() &&
          !isInsideModificator() &&
          isRootComponentBlockStarts(symbol)
        ) {
          openBracket();
          startComponent();
          // continue;
        } else if (symbol === '{') {
          if (isInsideComponent()) {
            openBracket();

            if (openedBrackets !== 1) {
              save(symbol);
            }
          } else if (isInsideModificator()) {
            openModificatorBracket();

            if (openedModificatorBrackets !== 1) {
              save(symbol);
            }
          }
        } else if (symbol === '}') {
          if (isInsideComponent()) {
            closeBracket();
          } else if (isInsideModificator()) {
            closeModificatorBracket();
          }

          if (isInsideComponent()) {
            if (isAllBracketsClosed()) {
              saveComponent();
            } else {
              save(symbol);
            }
          } else if (isInsideModificator()) {
            if (isAllModificatorBracketsClosed()) {
              saveModificator();
              clearAndRestore();
            } else {
              save(symbol);
            }
          }
        } else {
          save(symbol);
        }
      }

      restore();
      updateStylesTail();

      styles = styles.trim();

      if (styles || mixins.length) {
        precss.push({
          selector: parent.selector,
          media: parent.media,
          root: parent.root,
          styles: styles ? cssToJSON(styles, context, plugins) : {},
          mixins,
        });
      }

      return rootComponents;
    }
  };
}
