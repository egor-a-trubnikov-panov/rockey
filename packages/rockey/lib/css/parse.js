// const validateRule = (rule, value) => {};
// const autoprefixRule = (rule, value) => {};

const cache = new Map();

const parseCss = styles => {
  const stylesParts = styles
    .trim()
    .split(';')
    .map(s => s ? s.split(':') : null)
    .filter(s => !!s);

  const key = stylesParts.toString();

  if (!key.trim()) {
    return {};
  }

  if (cache.has(key)) {
    return cache.get(key);
  }

  const stylesJson = stylesParts.reduce(
    (json, value) => {
      if (value) {
        const key = value[0].trim();
        if (json[key]) {
          // css fallback values
          let prev = Array.isArray(json[key]) ? json[key] : [json[key]];

          json[value[0].trim()] = prev.concat([
            value.slice(1).join(':').trim(),
          ]);
        } else {
          json[value[0].trim()] = value.slice(1).join(':').trim();
        }

        // validateRule(value[0], json[value[0]]);
        // json[value[0].trim()] = autoprefixRule(value.slice(1).join(':').trim());
      }
      return json;
    },
    {}
  );

  cache.set(key, stylesJson);

  return stylesJson;
};

// @replace-start
const isModificatorStartSymbol = symbol => {
  return symbol === '@' || symbol === ':' || symbol === '[';
};
// @replace-end

// @replace-start
const isModificatorEndSymbol = symbol => {
  return symbol === '{' || symbol === ';';
};
// @replace-end

const isStartsWithModificator = raw => {
  return 0 === raw.indexOf('@media') ||
    0 === raw.indexOf('@keyframes') ||
    0 === raw.indexOf('[disabled]') ||
    0 === raw.indexOf(':focus') ||
    0 === raw.indexOf(':before') ||
    0 === raw.indexOf(':after') ||
    0 === raw.indexOf(':disabled') ||
    0 === raw.indexOf(':hover') ||
    0 === raw.indexOf(':link') ||
    0 === raw.indexOf(':active') ||
    0 === raw.indexOf(':visited') ||
    0 === raw.indexOf(':nth') ||
    0 === raw.indexOf('::first-line') ||
    0 === raw.indexOf(':not');
};

const parse = (raw, parent) => {
  let openedBrackets = 0;
  let openedModificatorBrackets = 0;
  let styles = '';

  let component = null;
  let components = {};

  let possibleModificator = 0;
  let modificators = {};
  let modificator = '';

  let current = '';
  let currentBackup = '';

  // @replace-start
  const restore = () => {
    current = currentBackup + current;
    currentBackup = '';
  };
  // @replace-end

  // @replace-start
  const clearAndRestore = () => {
    current = currentBackup;
    currentBackup = '';
  };
  // @replace-end

  // @replace-start
  const backup = () => {
    currentBackup = current;
    current = '';
  };
  // @replace-end

  // @replace-start
  const possibleModificatorStart = () => {
    possibleModificator = true;
  };
  // @replace-end

  // @replace-start
  const isPossibleModificator = () => {
    return possibleModificator;
  };
  // @replace-end

  // @replace-start
  const possibleModificatorEnd = () => {
    possibleModificator = false;
  };
  // @replace-end

  // @replace-start
  const isModificator = () => {
    return isStartsWithModificator(current.trim());
  };
  // @replace-end

  // @replace-start
  const openBracket = () => {
    openedBrackets++;
  };
  // @replace-end

  // @replace-start
  const closeBracket = () => {
    openedBrackets--;
  };
  // @replace-end

  // @replace-start
  const isAllBracketsClosed = () => {
    return openedBrackets === 0;
  };
  // @replace-end

  // @replace-start
  const openModificatorBracket = () => {
    openedModificatorBrackets++;
  };
  // @replace-end

  // @replace-start
  const closeModificatorBracket = () => {
    openedModificatorBrackets--;
  };
  // @replace-end

  // @replace-start
  const isAllModificatorBracketsClosed = () => {
    return openedModificatorBrackets === 0;
  };
  // @replace-end

  // @replace-start
  const startModificator = () => {
    modificator = current;
    current = '';
  };
  // @replace-end

  // @replace-start
  const saveModificator = () => {
    modificator = modificator.trim();

    // @remove-start
    if (modificators[modificator]) {
      throw new Error(
        `(parse error) "${modificator}" duplicated definiton at one block`
      );
    }
    // @remove-end

    modificators[modificator] = parse(current, {
      parentType: 'modificator',
      name: modificator,
    });

    modificator = '';
  };
  // @replace-end

  // @replace-start
  const isInsideModificator = () => {
    return !!modificator;
  };
  // @replace-end

  // @replace-start
  const isRootComponentBlockStarts = symbol => {
    return symbol === '{' && isAllBracketsClosed();
  };
  // @replace-end

  // @replace-start
  const startComponent = () => {
    const parts = current.trim().split(' ');

    if (parts.length === 1) {
      component = parts[0];
    } else {
      styles += parts.slice(0, -1).join(' ');
      component = parts[parts.length - 1];
    }

    current = '';
  };
  // @replace-end

  // @replace-start
  const isInsideComponent = () => {
    return !!component;
  };
  // @replace-end

  // @replace-start
  const save = symbol => {
    current += symbol;
  };
  // @replace-end

  // @replace-start
  const saveComponent = () => {
    // @remove-start
    if (components[component]) {
      throw new Error(
        `(parse error) "${component}" duplicated definiton at one block`
      );
    }
    // @remove-end

    components[component] = parse(current, {
      parentType: 'component',
      name: component,
    });

    component = null;
    current = '';
  };
  // @replace-end

  // @replace-start
  const updateStylesTail = () => {
    if (current) {
      styles += current;
    }
  };
  // @replace-end

  // ---
  const mixins = [];
  let possibleMixin = false;
  let mixin = '';

  // @replace-start
  const isMixinStartSymbol = symbol => {
    return symbol === '_';
  };
  // @replace-end

  // @replace-start
  const isMixinEndSymbol = symbol => {
    return symbol === ' ';
  };
  // @replace-end

  // @replace-start
  const possibleMixinStart = () => {
    possibleMixin = true;
  };
  // @replace-end

  // @replace-start
  const possibleMixinEnd = () => {
    possibleMixin = false;
  };
  // @replace-end

  // @replace-start
  const isPossibleMixin = () => {
    return possibleMixin;
  };
  // @replace-end

  // @replace-start
  const saveMixinString = symbol => {
    mixin += symbol;
  };
  // @replace-end

  // @replace-start
  const clearMixin = () => {
    mixin = '';
  };
  // @replace-end

  // @replace-start
  const removeMixinFromStyles = () => {
    current = current.replace(mixin.trim(), '');
  };
  // @replace-end

  // @replace-start
  const saveMixin = () => {
    mixins.push(mixin.trim());
  };
  // @replace-end

  // @remove-start
  const isInValidMixinPosition = () => {
    return !parent ||
      (currentBackup.trim().length &&
        !new RegExp(`(;|{)\\s+?${mixin.trim()}`).test(`${current}`.trim()));
  };
  // @remove-end

  // ----

  for (let symbol of raw) {
    if (!isInsideComponent() && !isInsideModificator()) {
      if (isMixinStartSymbol(symbol)) {
        possibleMixinStart();
      }

      if (isPossibleMixin()) {
        saveMixinString(symbol);
        if (isMixinEndSymbol(symbol)) {
          // TODO:
          // if (isMixin()) {
          //
          // }

          possibleMixinEnd();
          // @remove-start
          if (isInValidMixinPosition()) {
            if (parent) {
              const nearestCode = `${currentBackup}${current}`
                .replace(/\s+/g, ' ')
                // .replace(mixin.trim(), '{{ MIXIN }}' + mixin)
                .trim();

              console.warn(
                'seems wrong mixin position',
                `${parent.name} - ${nearestCode}`
              );
            } else {
              console.warn(
                'seems wrong mixin position',
                `mixin should not be at root`
              );
            }
          }
          // @remove-end

          saveMixin();
          removeMixinFromStyles();
          clearMixin();
        }
      }
    }

    if (!isInsideComponent() && !isInsideModificator()) {
      if (!isPossibleModificator() && isModificatorStartSymbol(symbol)) {
        possibleModificatorStart();
        backup();
      }

      if (isPossibleModificator() && isModificatorEndSymbol(symbol)) {
        if (isModificator()) {
          startModificator();
        } else {
          restore();
        }

        possibleModificatorEnd();
      }
    }

    if (
      !isPossibleModificator() &&
      !isInsideModificator() &&
      isRootComponentBlockStarts(symbol)
    ) {
      openBracket();
      startComponent();
      // continue;
    } else if (symbol === '{') {
      if (isInsideComponent()) {
        openBracket();

        if (openedBrackets !== 1) {
          save(symbol);
        }
      } else if (isInsideModificator()) {
        openModificatorBracket();

        if (openedModificatorBrackets !== 1) {
          save(symbol);
        }
      }
    } else if (symbol === '}') {
      if (isInsideComponent()) {
        closeBracket();
      } else if (isInsideModificator()) {
        closeModificatorBracket();
      }

      if (isInsideComponent()) {
        if (isAllBracketsClosed()) {
          saveComponent();
        } else {
          save(symbol);
        }
      } else if (isInsideModificator()) {
        if (isAllModificatorBracketsClosed()) {
          saveModificator();
          clearAndRestore();
        } else {
          save(symbol);
        }
      }
    } else {
      save(symbol);
    }
  }

  restore();
  updateStylesTail();

  return {
    mixins,
    components,
    styles: parseCss(styles),
    modificators,
  };
};

export default inline => {
  return parse(inline.replace(/\r|\n/g, ''));
};
