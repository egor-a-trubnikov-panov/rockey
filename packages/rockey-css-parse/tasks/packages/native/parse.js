import { argv } from 'yargs';
import chalk from 'chalk';

import getNativeCss from './utils/getCss';
import timer from './utils/timer';
import output from './utils/output';

import createParser from '../../../lib/parse';
import { stringifyRules } from '../../../lib/stringify';

const size = argv.size || 1000;
const css = getNativeCss(size);

console.log('');
console.log(
  `Start ${chalk.green('parse')} measuring with ${chalk.cyan(size)} class length`
);
console.log('');

const parse = createParser();

timer(() => {
  stringifyRules(parse(css).precss);
}).then(time => {
  output('parse', size, time);
});
