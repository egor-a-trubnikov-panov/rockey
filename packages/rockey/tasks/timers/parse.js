import { argv } from 'yargs';
import chalk from 'chalk';

import getCss from './utils/getCss';
import timer from './utils/timer';
import output from './utils/output';

import parse from '../../lib/css/parse';
import { stringifyRules } from '../../lib/styleSheets/utils/stringify';

const size = argv.size || 10;
const css = getCss(size);

console.log('');
console.log(
  `Start ${chalk.green('parse')} measuring with ${chalk.cyan(size)} class length`
);
console.log('');

timer(() => {
  stringifyRules(parse(css).precss);
}).then(time => {
  output('parse', size, time);
});
