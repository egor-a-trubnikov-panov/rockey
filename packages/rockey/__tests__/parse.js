import parse from '../lib/css/parse';

export const styles = {
  empty: '',
  //-----
  onlyStyles: `
    color: white;
    font-weight: bold;
  `,
  //-----
  withComponents: `
    Button {
      color: red;
    }
  `,
  //-----
  withNestedComponents: `
    Button {
      color: red;

      Icon {
        color: blue;
      }
    }
  `,
  //-----
  withModificators: `
    Button {
      color: red;

      :hover {
        color: purple;
      }

      ::before {
        width: 16px;
      }

      Icon {
        color: blue;
      }
    }
  `,
  //-----
  withComponentsInsideModificators: `
    Button {
      color: red;

      :hover {
        color: purple;

        Icon {
          color: blue;
        }
      }
    }
  `,

  //-----
  withModificatorsInsideComponents: `
    Button {
      color: red;

      Icon {
        color: blue;

        :hover {
          color: purple;
        }
      }
    }
  `,

  //-----
  withMedia: `
    Button {
      color: red;

      Icon {
        color: blue;

        @media (max-width: 699px) {
          background: green;

          :hover {
            color: purple;
          }
        }

        :hover {
          color: purple;
        }
      }

      @media screen and (max-width: 699px) and (min-width: 520px) {
        background: yellow;
      }
    }
  `,
  // ---
  multipleMediaAtOneLevel: `
    Button {
      @media (max-width: 699px) {
        background: green;
        :hover {
          color: green;
        }
      }
    }

    Button2 {
      @media (max-width: 699px) {
        background: blue;
        :hover {
          color: blue;
        }
      }
    }
  `,

  // ---
  withFallbackValues: `
    Button {
      color: rgba(0, 255, 0, 0.3);
      color: #00ff00;
    }
  `,

  // ---
  withKeyframes: `
    Button {
      @keyframes example-first {
        0% { color: red; }
        50% { color: yellow; }
        100% { color: red; }
      }

      animation: example-first 1s infinity;
    }

    Button2 {
      @keyframes example-second {
        0% { color: purple; }
        100% { color: black; }
      }

      animation-name: example-second;
      animation-iteration-count: infinity;
      animation-duraiton: 1s;
    }
  `,

  // ---
  withTagSelectors: `
    Button {
      color: red;

      span.text {
        color: black;
      }

      div#hardcode {
        color: white;
      }
    }
  `,

  // ---
  withMostModificators: `
    Button {
      color: red;

      ::first-line {
        background-color: yellow;
      }

      :disabled {
        color: green;
      }

      :hover {
        color: blue;
      }

      ::after {
        color: purple;
      }

      :nth-child(2) {
        color: white;
      }

      :not(.hardcodeClass) {
        color: yellow;
      }

      :not(Icon) {
        color: orange;
      }
    }
  `,
};

export const struct = {
  mixins: [],
  components: {},
  styles: {},
  modificators: {},
  combinedComponents: [],
};

describe('parse', () => {
  it('empty', () => {
    const parsed = parse(styles.empty);
    expect(parsed).toEqual(struct);
  });

  it('onlyStyles', () => {
    const parsed = parse(styles.onlyStyles);

    expect(parsed).toEqual({
      ...struct,
      styles: {
        color: 'white',
        'font-weight': 'bold',
      },
    });
  });

  it('withComponents', () => {
    const parsed = parse(styles.withComponents);

    expect(parsed).toEqual({
      ...struct,
      components: {
        Button: {
          ...struct,
          styles: {
            color: 'red',
          },
        },
      },
    });
  });

  it('withNestedComponents', () => {
    const parsed = parse(styles.withNestedComponents);

    expect(parsed).toEqual({
      ...struct,
      components: {
        Button: {
          ...struct,
          styles: {
            color: 'red',
          },
          components: {
            Icon: {
              ...struct,
              styles: {
                color: 'blue',
              },
            },
          },
        },
      },
    });
  });

  it('withModificators', () => {
    const parsed = parse(styles.withModificators);

    expect(parsed).toEqual({
      ...struct,
      components: {
        Button: {
          ...struct,
          modificators: {
            ':hover': {
              ...struct,
              styles: {
                color: 'purple',
              },
            },
            '::before': {
              ...struct,
              styles: {
                width: '16px',
              },
            },
          },
          styles: {
            color: 'red',
          },
          components: {
            Icon: {
              ...struct,
              styles: {
                color: 'blue',
              },
            },
          },
        },
      },
    });
  });

  it('withComponentsInsideModificators', () => {
    const parsed = parse(styles.withComponentsInsideModificators);

    expect(parsed).toEqual({
      ...struct,
      components: {
        Button: {
          ...struct,
          modificators: {
            ':hover': {
              ...struct,
              components: {
                Icon: {
                  ...struct,
                  styles: {
                    color: 'blue',
                  },
                },
              },
              styles: {
                color: 'purple',
              },
            },
          },
          styles: {
            color: 'red',
          },
        },
      },
    });
  });

  it('withModificatorsInsideComponents', () => {
    const parsed = parse(styles.withModificatorsInsideComponents);

    expect(parsed).toEqual({
      ...struct,
      components: {
        Button: {
          ...struct,
          styles: {
            color: 'red',
          },
          components: {
            Icon: {
              ...struct,
              modificators: {
                ':hover': {
                  ...struct,
                  styles: {
                    color: 'purple',
                  },
                },
              },
              styles: {
                color: 'blue',
              },
            },
          },
        },
      },
    });
  });

  it('withMedia', () => {
    const parsed = parse(styles.withMedia);

    expect(parsed).toEqual({
      ...struct,
      components: {
        Button: {
          ...struct,
          components: {
            Icon: {
              ...struct,
              styles: {
                color: 'blue',
              },
              modificators: {
                '@media (max-width: 699px)': {
                  ...struct,
                  styles: {
                    background: 'green',
                  },
                  modificators: {
                    ':hover': {
                      ...struct,
                      styles: {
                        color: 'purple',
                      },
                    },
                  },
                },
                ':hover': {
                  ...struct,
                  styles: {
                    color: 'purple',
                  },
                },
              },
            },
          },
          styles: {
            color: 'red',
          },
          modificators: {
            '@media screen and (max-width: 699px) and (min-width: 520px)': {
              ...struct,
              styles: {
                background: 'yellow',
              },
            },
          },
        },
      },
    });
  });

  it('multipleMediaAtOneLevel', () => {
    const parsed = parse(styles.multipleMediaAtOneLevel);

    expect(parsed).toEqual({
      ...struct,
      components: {
        Button: {
          ...struct,
          modificators: {
            '@media (max-width: 699px)': {
              ...struct,
              modificators: {
                ':hover': {
                  ...struct,
                  styles: {
                    color: 'green',
                  },
                },
              },
              styles: {
                background: 'green',
              },
            },
          },
        },
        Button2: {
          ...struct,
          modificators: {
            '@media (max-width: 699px)': {
              ...struct,
              modificators: {
                ':hover': {
                  ...struct,
                  styles: {
                    color: 'blue',
                  },
                },
              },
              styles: {
                background: 'blue',
              },
            },
          },
        },
      },
    });
  });

  it('withFallbackValues', () => {
    const parsed = parse(styles.withFallbackValues);

    expect(parsed).toEqual({
      ...struct,
      components: {
        Button: {
          ...struct,
          styles: {
            color: ['rgba(0, 255, 0, 0.3)', '#00ff00'],
          },
        },
      },
    });
  });

  it('withKeyframes', () => {
    const parsed = parse(styles.withKeyframes);

    expect(parsed).toEqual({
      ...struct,
      components: {
        Button: {
          ...struct,
          modificators: {
            '@keyframes example-first': {
              ...struct,
              components: {
                '0%': {
                  ...struct,
                  styles: {
                    color: 'red',
                  },
                },
                '50%': {
                  ...struct,
                  styles: {
                    color: 'yellow',
                  },
                },
                '100%': {
                  ...struct,
                  styles: {
                    color: 'red',
                  },
                },
              },
            },
          },
          styles: {
            animation: 'example-first 1s infinity',
          },
        },
        Button2: {
          ...struct,
          modificators: {
            '@keyframes example-second': {
              ...struct,
              components: {
                '0%': {
                  ...struct,
                  styles: {
                    color: 'purple',
                  },
                },
                '100%': {
                  ...struct,
                  styles: {
                    color: 'black',
                  },
                },
              },
            },
          },
          styles: {
            'animation-name': 'example-second',
            'animation-iteration-count': 'infinity',
            'animation-duraiton': '1s',
          },
        },
      },
    });
  });

  it('withTagSelectors', () => {
    const parsed = parse(styles.withTagSelectors);

    expect(parsed).toEqual({
      ...struct,
      components: {
        Button: {
          ...struct,
          styles: {
            color: 'red',
          },
          components: {
            'span.text': {
              ...struct,
              styles: {
                color: 'black',
              },
            },
            'div#hardcode': {
              ...struct,
              styles: {
                color: 'white',
              },
            },
          },
        },
      },
    });
  });

  it('withMostModificators', () => {
    const parsed = parse(styles.withMostModificators);

    expect(parsed).toEqual({
      ...struct,
      components: {
        Button: {
          ...struct,
          styles: {
            color: 'red',
          },
          modificators: {
            '::first-line': {
              ...struct,
              styles: {
                'background-color': 'yellow',
              },
            },
            ':disabled': {
              ...struct,
              styles: {
                color: 'green',
              },
            },
            ':hover': {
              ...struct,
              styles: {
                color: 'blue',
              },
            },
            '::after': {
              ...struct,
              styles: {
                color: 'purple',
              },
            },
            ':nth-child(2)': {
              ...struct,
              styles: {
                color: 'white',
              },
            },
            ':not(.hardcodeClass)': {
              ...struct,
              styles: {
                color: 'yellow',
              },
            },
            ':not(Icon)': {
              ...struct,
              styles: {
                color: 'orange',
              },
            },
          },
        },
      },
    });
  });
});

test('parse: customStyles', () => {
  const parsed = parse(
    `
    Button {
      box-shadow: 0 2px 5px rgba(0, 0, 0, .26);

      Icon {
        color: #fff
      }

      margin: 5px 6px;
    }
  `
  );

  expect(parsed).toMatchSnapshot();
});
