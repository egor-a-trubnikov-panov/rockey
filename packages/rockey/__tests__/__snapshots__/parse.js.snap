// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse: customStyles 1`] = `
Object {
  "components": Object {
    "Button": Object {
      "components": Object {
        "Icon": Object {
          "components": Object {},
          "mixins": Array [],
          "modificators": Object {},
          "styles": Object {
            "color": "#fff",
          },
        },
      },
      "mixins": Array [],
      "modificators": Object {},
      "styles": Object {
        "box-shadow": "0 2px 5px rgba(0, 0, 0, .26)",
        "margin": "5px 6px",
      },
    },
  },
  "mixins": Array [],
  "modificators": Object {},
  "styles": Object {},
}
`;
